hackforla/VRMS
# Fix unit tests - Backend

Goal: Get the tests working

The tests have been disabled for some time and there seems to be a growing number of regressions when pulling up the application. The goal of this projects was to get the tests working, which largely centered on setting up test frameworks for the development environment.

Work Done:
- The backend was split into it's own directory. This was a requirement to get the test runners working in the backend. The errors were raised by two different test runners. 
- Implement a universal unit test runner. As the project is largely a react app, I aimed to get use the same test runner that is being used in the front end. Jest requires some additional and nuanced setup for backend Node purposes, which is largely caught in in-memory-mongodb setup. 
- Get working unit tests and write unit test examples
- Get working integration tests for the API and write examples
- Get documentation for developers to use their local database
- Make the database for the automated tests be idempotent and not relate to their local development environments. 

How to test that this work?
1. Pull my branch locally
 - git remote add nickbeaird GIT REPO LINK
 - git fetch nickbeaird
 - git checkout nickbeaird/add_frontend_testing

* Note that you now have a backend directory

2. Add your .env file at the root directory to the backend directory

3. Delete all of your package.json file and nod_modules directories
4. Run npm install in the root, backend, and client directories. 
 - cd root && npm install 
 - cd client && npm install 
 - cd ../backend && npm install
7. Navigate to the backend directory and run the tests
 - npm run test (see package.json)
